src/data/nat/digits.lean

```diff
@@ -0,0 +1,376 @@
+end
+
+-- This is really a theorem about polynomials.
+lemma dvd_of_digits_sub_of_digits {α : Type*} [comm_ring α]
+  {a b k : α} (h : k ∣ a - b) (L : list ℕ) :
+  k ∣ of_digits a L - of_digits b L :=
+begin
+  induction L with d L ih,
+  { change k ∣ 0 - 0, simp, },
+  { simp only [of_digits, add_sub_add_left_eq_sub],
+    exact dvd_mul_sub_mul h ih, }
+end
+
+lemma of_digits_modeq' (b b' : ℕ) (k : ℕ) (h : b ≡ b' [MOD k]) (L : list ℕ) :
+  of_digits b L ≡ of_digits b' L [MOD k] :=
+begin
+  induction L with d L ih,
+  { refl, },
+  { dsimp [of_digits],
+    dsimp [nat.modeq] at *,
+    conv_lhs { rw [nat.add_mod, nat.mul_mod, h, ih], },
+    conv_rhs { rw [nat.add_mod, nat.mul_mod], }, }
+end
+
+lemma of_digits_modeq (b k : ℕ) (L : list ℕ) : of_digits b L ≡ of_digits (b % k) L [MOD k] :=
+of_digits_modeq' b (b % k) k (nat.modeq.symm (nat.modeq.mod_modeq b k)) L
+
+lemma of_digits_mod (b k : ℕ) (L : list ℕ) : of_digits b L % k = of_digits (b % k) L % k :=
+of_digits_modeq b k L
+
+lemma of_digits_zmodeq' (b b' : ℤ) (k : ℕ) (h : b ≡ b' [ZMOD k]) (L : list ℕ) :
+  of_digits b L ≡ of_digits b' L [ZMOD k] :=
+begin
+  induction L with d L ih,
+  { refl, },
+  { dsimp [of_digits],
+    dsimp [int.modeq] at *,
+    conv_lhs { rw [int.add_mod, int.mul_mod, h, ih], },
+    conv_rhs { rw [int.add_mod, int.mul_mod], }, }
+end
+
+lemma of_digits_zmodeq (b : ℤ) (k : ℕ) (L : list ℕ) :
+  of_digits b L ≡ of_digits (b % k) L [ZMOD k] :=
+of_digits_zmodeq' b (b % k) k (int.modeq.symm (int.modeq.mod_modeq b ↑k)) L
+
+lemma of_digits_zmod (b : ℤ) (k : ℕ) (L : list ℕ) :
+  of_digits b L % k = of_digits (b % k) L % k :=
+of_digits_zmodeq b k L
+
+lemma modeq_digits_sum (b b' : ℕ) (h : b' % b = 1) (n : ℕ) :
+   n ≡ (digits b' n).sum [MOD b] :=
+begin
+  rw ←of_digits_one,
+  conv { congr, skip, rw ←(of_digits_digits b' n) },
+  convert of_digits_modeq _ _ _,
+  exact h.symm,
+end
+
+lemma modeq_three_digits_sum (n : ℕ) : n ≡ (digits 10 n).sum [MOD 3] :=
+modeq_digits_sum 3 10 (by norm_num) n
+
+lemma modeq_nine_digits_sum (n : ℕ) : n ≡ (digits 10 n).sum [MOD 9] :=
+modeq_digits_sum 9 10 (by norm_num) n
+
+lemma dvd_iff_dvd_digits_sum (b b' : ℕ) (h : b' % b = 1) (n : ℕ) :
+  b ∣ n ↔ b ∣ (digits b' n).sum :=
+begin
+  rw ←of_digits_one,
+  conv_lhs { rw ←(of_digits_digits b' n) },
+  rw [nat.dvd_iff_mod_eq_zero, nat.dvd_iff_mod_eq_zero, of_digits_mod, h],
+end
+
+lemma three_dvd_iff (n : ℕ) : 3 ∣ n ↔ 3 ∣ (digits 10 n).sum :=
+dvd_iff_dvd_digits_sum 3 10 (by norm_num) n
+
+lemma nine_dvd_iff (n : ℕ) : 9 ∣ n ↔ 9 ∣ (digits 10 n).sum :=
+dvd_iff_dvd_digits_sum 9 10 (by norm_num) n
+
+lemma zmodeq_of_digits_digits (b b' : ℕ) (c : ℤ) (h : b' ≡ c [ZMOD b]) (n : ℕ) :
+  n ≡ of_digits c (digits b' n) [ZMOD b] :=
+begin
+  conv { congr, skip, rw ←(of_digits_digits b' n) },
+  rw coe_int_of_digits,
+  apply of_digits_zmodeq' _ _ _ h,
+end
+
+namespace list
+
+/-- The alternating sum of elements in a list. -/
+-- As we don't prove anything about this yet,
```

Comment:
This is now no longer true. So we should probably start moving things and prove the basic supporting API. I will probably have a bit of time for this in the next two days.

Reply:
Done